tests/integration/test_mongodb.py

```diff
@@ -140,72 +121,8 @@ async def test_empty_collections(mongodb: MongoDbFixture):
     assert db.list_collection_names() == []
 
 
-async def test_dao_happy(mongodb: MongoDbFixture):
-    """Test the happy path of performing basic CRUD database interactions using
-    the MongoDbDaoFactory in a surrogate ID setting.
-    """
-    dao = await mongodb.dao_factory.get_dao(
-        name="example",
-        dto_model=ExampleDto,
-        dto_creation_model=ExampleCreationDto,
-        id_field="id",
-    )
-
-    # insert an example resource:
-    resource_to_create = ExampleCreationDto(field_a="test1", field_b=27, field_c=True)
-    resource_inserted = await dao.insert(resource_to_create)
-
-    assert isinstance(resource_inserted, ExampleDto)
-    assert (
-        resource_inserted.model_dump(exclude={"id"}) == resource_to_create.model_dump()
-    )
-
-    # read the newly inserted resource:
-    resource_read = await dao.get_by_id(resource_inserted.id)
-
-    assert resource_read == resource_inserted
-
-    # update the resource:
-    resource_update = resource_inserted.model_copy(update={"field_c": False})
-    await dao.update(resource_update)
-
-    # read the updated resource again:
-    resource_updated = await dao.get_by_id(resource_inserted.id)
-
-    assert resource_update == resource_updated
-
-    # insert additional resources:
-    add_resources_to_create = (
-        ExampleCreationDto(field_a="test2", field_b=27, field_c=True),
-        ExampleCreationDto(field_a="test3", field_b=27, field_c=False),
-    )
-    add_resources_inserted = [
-        await dao.insert(resource) for resource in add_resources_to_create
-    ]
-
-    # perform a search for multiple resources:
-    obtained_hits = {
-        hit async for hit in dao.find_all(mapping={"field_b": 27, "field_c": False})
-    }
-
-    expected_hits = {resource_updated, add_resources_inserted[1]}
-    assert obtained_hits == expected_hits
-
-    # find a single resource:
-    obtained_hit = await dao.find_one(mapping={"field_a": "test1"})
-
-    assert obtained_hit == resource_updated
-
-    # delete the resource:
-    await dao.delete(resource_inserted.id)
-
-    # confirm that the resource was deleted:
-    with pytest.raises(ResourceNotFoundError):
-        _ = await dao.get_by_id(resource_inserted.id)
-
-
-async def test_dao_insert_natural_id_happy(mongodb: MongoDbFixture):
-    """Tests the happy path of inserting a new resource in a natural ID setting."""
+async def test_dao_insert_happy(mongodb: MongoDbFixture):
```

Comment:
I think we also need to test the `insert_unhappy` behavior (e.g. when the id already exists).

Reply:
I would add the happy CRUD test as well even when the individual steps are already tested.

The CRUD test is also useful as a compact documentation how to use the DAO.

Reply:
The duplicate ID behavior is already tested at the end of that test. After the successful insertion, it attempts a repeat and looks for the error. I will move it into a separate test though.

Reply:
Yes, either remove the "happy" from the name or (better) move it to a separate test.

Reply:
Please see [8036561](https://github.com/ghga-de/hexkit/pull/127/commits/80365614ba7c94bec114837add7cdb9bcf901c06) for the update

Reply:
I wonder if we should add a similar, but separate test where we mock uuid.hex to produce duplicate values (like 1, 1, 2, 2 , ...) to simulate the (extremely unlikely) case that a duplicate UUID is produced. The test could then also document what happens in this case and how it can be recovered (simple retry). Maybe the insert method could have a parameter `retries_if_duplicate`?

Reply:
Considering ID generation happens before/outside of `insert()`, I feel the responsibility of the interpreting and handling this case falls on the user. I don't think there's a way for a UUID collision to be detected by a the DAO, unless we're talking about different things.

Reply:
That's true. The insert method cannot just retry, but would need to do something like this to recreate the id before trying to insert again;

```python
dto = Dto(**dto.model_dump(exclude={self._id_field}))
```

Reply:
But the problem is then that the user would not be aware that the id has changed since the dto is not passed back.

Reply:
Probably this is too much magic, and we should leave the responsibility for handling this to the user as you said.

But it would still be good to have a test as example how to do handle this case.

Reply:
I would still add a separate test case that shows what happens in the case where UUID4Field creates a duplicate value (can be simulated by mocking uuid4) and how the user could handle this case (by recreating the object excluding the id field and retrying). Again, mainly meant as living documentation.